### DISCUSSION TITLE
How many scripts can be run safely with one invocation of bpftrace?

### DISCUSSION THREAD

**Discussion by Engg-blogger** [bpftrace/bpftrace#3273] [General]
Hi - I am running a bt script with bpftrace that is attached to multiple krpobes (about 4). I am planning to add one more tracepoint to the same file (as we have limited memory constraints and cannot invoke multiple instances of bpftrace). Is there any guidance on how many scripts will be too much for bpftrace to handle at runtime? Or any other constraint i should be aware of?
Is there any other way where i can run multiple kprobes with single bpftrace invocation and output the data in different files?

**Comment by jordalgo** [comment]
> Is there any guidance on how many scripts will be too much for bpftrace to handle at runtime? 

Right now you can only run a single script at a time e.g. `sudo bpftrace myscript.bt` (though multiple invocations works). Do you mean probes inside a single script? If so, there is currently no limitation. That being said attaching to 100s of probes might take a while and also, depending on how much logic is in the probe program, there might be some impact on kernel perf. 

> Is there any other way where i can run multiple kprobes with single bpftrace invocation and output the data in different files?

No, that's currently not possible but it's an interesting use case. Would this just be to get around multiple invocations of bpftrace?

**Reply by Engg-blogger** [reply]
Yes.. Each invocation of bpftrace consumes some memory and if we have to deal with low memory VM's, this becomes a challenge. So I was thinking if there is a way to run bpftrace as a service and invoke multiple scripts and pipe each scripts output separately. That looks like a good ask for scalability perspective as well. Would like to know if this is something which can be achieved or a workaround for this?

**Reply by jordalgo** [reply]
Unfortunately, I don't think that's currently possible. The internals of bpftrace are setup to handle one script at a time. That being said, maybe AOT (even though it's in beta testing right now) might be a solution for this as you'd compile all your scripts before hand and that might save resources when you actually run them. Feel free to create an issue for this feature request and we can triage.

**Reply by ajor** [reply]
@the-prachi What about doing something like this:
```
kprobe:my_first_script { $script_id = "first"; print(($script_id, $regular_data_to_print)); }
kprobe:my_second_script { $script_id = "second"; print(($script_id, $regular_data_to_print)); }
```

If you use the JSON output format (`-f json`) then it should be easy to programatically extract the script ID from each line of output and separate the data as a separate step.

**Reply by Engg-blogger** [reply]
This makes sense. I was thinking of doing the same with stdout and appending script name in print statement. Thats why i was worried if 5-6 kprobes could be overkill for bpftrace.

**Reply by ajor** [reply]
Cool, 5-6 should be no problem at all (it's pretty normal in fact)
